### ISSUE TITLE
[css-text-decor-4] Consider renaming `text-decoration-trim`

### ISSUE THREAD

**Issue by SebastianZ** [w3c/csswg-drafts#8402]
Besides the general discussion in #4557 to introduce a way to modify the length of the decoration, there was also some discussion around how to call the property. As I [previously summarized](https://github.com/w3c/csswg-drafts/issues/4557#issuecomment-1184944512), name suggestions for it so far were:

* `text-decoration-trim`
* `text-decoration-inset`
* `text-decoration-limit`
* `text-decoration-clip`

As there are also use cases for _extending_ the decorations beyond their normal size. In my opinion, "trim" doesn't express that well enough. Therefore I slightly tend towards "limit".

We have a poll running in comments below:
https://github.com/w3c/csswg-drafts/issues/8402#issuecomment-3408427694
https://github.com/w3c/csswg-drafts/issues/8402#issuecomment-3408451101

Sebastian

**Comment by jfkthame** [comment]
We're in the process of implementing this, so if there's going to be some naming bikeshedding, we'd appreciate it happening sooner rather than later.

(Personally, I'm happy with `text-decoration-trim`; the idea that "negative trimming" extends the line doesn't bother me. Neither `-limit` nor `-clip` seem nearly as clear; `-inset` would be OK but no better than `-trim`.)

**Comment by SebastianZ** [comment]
I'd argue that `*-inset` is definitely better than `*-trim` and would also be more conform with other properties like `view-timeline-inset` or `inset-*`.

All the other properties that have "trim" in their name like `white-space-trim`, `text-box-trim`, `margin-trim`, etc. are keyword-based. And `trim` rather indicates cutting off something rather than potentially expanding it. Therefore I am strongly against using `*-trim` and `*-clip` and prefer either `*-inset` or `*-limit`.

Sebastian

**Comment by jfkthame** [comment]
Thinking about this again, I'm feeling less in favor of `-inset` here. Not sure how well I can articulate this, but fwiw....

To my mind, `inset` properties are about adjusting their target element _in relation to another element_ (a parent or containing box). Given an element that is being sized or placed within an area that has been defined by some external thing, it makes sense to have the ability to "inset" (or expand, by use of negative inset) that area.

But with decoration lines, it's not readily apparent to an author that there is any such "container" to apply insets to. The decoration line seems more like an attribute of a range of text, and its length is determined by the length of that text, not in relation to some containing element, so the idea of "insetting" from such a container is a little odd.

Arguably the inline element (or pseudo, e.g. `::selection`) that defines the decorating line could be considered that container, and the spec talks of the "decorating box", so insets can reasonably be applied to that box, but I don't think this is a model that will generally be at the front of an author's mind. To an author, a decorating line is an independent thing that simply appears below (or above) a range of text. The property under discussion here directly adjusts the length of that line, by "trimming" a certain amount from each end, not by "insetting" it from a containing area.

(As for negative values, a negative "trim" extends and a negative "inset" expands; these seem equally reasonable concepts to me.)

**Comment by SebastianZ** [comment]
Just to throw in some more alternatives:

* `text-decoration-adjust`
* `text-decoration-{size|length}-adjust`
* `text-decoration-resize`

Not necessarily saying that those are any better than the previous options.

> To my mind, `inset` properties are about adjusting their target element _in relation to another element_ (a parent or containing box). Given an element that is being sized or placed within an area that has been defined by some external thing, it makes sense to have the ability to "inset" (or expand, by use of negative inset) that area.
> 
> But with decoration lines, it's not readily apparent to an author that there is any such "container" to apply insets to. The decoration line seems more like an attribute of a range of text, and its length is determined by the length of that text, not in relation to some containing element, so the idea of "insetting" from such a container is a little odd.

"Insetting" to me means relative to _something_, not necessarily a containing element. Though I can see your point. The existing `*-inset-*` properties we have are also related to some containing element.

Having said that, I still believe "limit", "inset" and possibly also "(length-)adjust" express the meaning better than words in the meaning of cutting off or taking away something.

Though I am not a native speaker. And so far there were only a few opinions expressed. Maybe someone could set up a quick naming poll somewhere to reach a broader audience, maybe @mirisuzanne or @jensimmons? (I'm not on the big messaging/social media platforms, so can't do that myself, unfortunately.)

Sebastian

**Comment by mirisuzanne** [comment]
I don't have strong feelings here, `text-decoration-trim` seems ok to me. `*-edge` might also be an option?

I'm happy to share a poll – but Mastodon only allows 4 options, and Bluesky doesn't have built-in polling at all. We could try an emoji poll here instead, and link from social media posts?

**Comment by fantasai** [comment]
There's something similar here with the `column-rule` properties that @kbabbitt has been working to spec in https://www.w3.org/TR/css-gaps-1/#outset ... They're using `*-rule-outset`. Maybe `text-decoration-outset` would work here? Keeping them consistent seems like a good plan in any case.

**Comment by kbabbitt** [comment]
Gap Decorations does use outsets right now, though in #12848 I'm contemplating flipping them to insets.

For the property we're discussing here, which do we expect to be the more common use case - extending or shrinking? Depending on the answer to that, I'd prefer either `outset` or `inset` respectively, such that positive values would reflect the more common case. I would prefer either of those over `trim` for the reasons @SebastianZ gave. I don't like `adjust` because that's not clear on which is the positive-value direction, or even what's being adjusted - I could see it being confused for adjustment up or down from the text baseline, for example.

**Comment by SebastianZ** [comment]
> For the property we're discussing here, which do we expect to be the more common use case - extending or shrinking?

From my experience, shrinking is _much_ more common than extending. See the examples I linked to in #8403.

> I don't like `adjust` because that's not clear on which is the positive-value direction, or even what's being adjusted - I could see it being confused for adjustment up or down from the text baseline, for example.

Therefore, I also provided a `text-decoration-{size|length}-adjust` option. But I admit, it's a bit lengthy.

Sebastian

**Comment by SebastianZ** [comment]
As suggested by @mirisuzanne, here's the poll (options sorted by when they got suggested):

* 👍🏻 `text-decoration-trim`
* 😄 `text-decoration-inset`
* 🎉 `text-decoration-limit`
* ❤️ `text-decoration-clip`
* 🚀 `text-decoration-adjust`
* 👀 `text-decoration-size-adjust`

I'm splitting them into two comments, because there are not enough emojis.

Multiple votes are allowed!

@fantasai noted that the suffix we use for this property should also be used for the `{row|column}-rule-*` properties for consistence.

Let's see if we can get some results before the call.

Sebastian

PS: I switched around the icons before any votes were casted.

**Comment by SebastianZ** [comment]
Here's the second part of options:
 
* 👍🏻 `text-decoration-length-adjust`
* 😄 `text-decoration-resize`
* 🎉 `text-decoration-edge`

Sebastian

**Comment by davidleininger** [comment]
Not to add to the bike shedding, but what about `text-decoration-spread`? Similar to the way that we do box-shadow. 

- `*-spread: -5px;` make the line smaller.
- `*-spread: 5px;` makes the line wider.

**Comment by jfkthame** [comment]
> Not to add to the bike shedding, but what about `text-decoration-spread`? 

IMO "spread" sounds like it might thicken the line in all directions, kind of like a blur operation.

**Comment by Crissov** [comment]
`text-decoration-extent` would be positive to grow and negative to shrink from the default. 

PS: But really most options seem fine compared to some other naming falls.

**Comment by davidleininger** [comment]
> `text-decoration-extent`

@Crissov Do you mean `text-decoration-extend`?

> IMO "spread" sounds like it might thicken the line in all directions, kind of like a blur operation.

@jfkthame My thoughts were that `text-decoration-thickness` already being a property would help for making `text-decoration-spread` clear, but I hear that concern.

**Comment by SebastianZ** [comment]
> Not to add to the bike shedding, but what about `text-decoration-spread`? Similar to the way that we do box-shadow.

I also thought of `spread` at some point, though given most use cases want to _shrink_ the lines, I think it should rather be a word that means the opposite.

Sebastian

**Comment by Crissov** [comment]
@davidleininger No, since the noun _extent_ is relatively neutral, whereas the verb _extend_ is more associated with increase than decrease. However, your confusion is a good indicator for this term being a bad choice on my behalf: retracted. 

PS: `clip` and `inset` usually relate to 2D adjustments in other CSS properties and keyword values, not 1D.

**Comment by jfkthame** [comment]
> [@davidleininger](https://github.com/davidleininger) No, since the noun _extent_ is relatively neutral, whereas the verb _extend_ is striker associated with increase than decrease.

But _extent_ sounds like the property is defining the absolute _size_ (extent) of the decoration, not an _adjustment_ to its default size. So I don't think that works.

**Comment by jfkthame** [comment]
Re-adding the Agenda label, as it would be good to revisit this and make a firm decision sooner rather than later, given that implementation work is ongoing.

**Comment by SebastianZ** [comment]
As requested by @fantasai, here's a fresh poll (options sorted by when they got suggested):

* 👍🏻 `text-decoration-trim`
* 😄 `text-decoration-inset`
* 🎉 `text-decoration-limit`
* ❤️ `text-decoration-clip`
* 🚀 `text-decoration-adjust`
* 👀 `text-decoration-size-adjust`

I'm splitting them into two comments, because there are not enough emojis.

Multiple votes are allowed! Please read the comments for context! Note that we [resolved on](https://github.com/w3c/csswg-drafts/issues/8402#issuecomment-3407283833) renaming `text-decoration-*` and the `{row|column}-rule-*` properties for adjusting their length to a common suffix. So when choosing your option please keep that in mind!

Sebastian

**Comment by SebastianZ** [comment]
And here's the second part of options:
 
* 👍🏻 `text-decoration-length-adjust`
* 😄 `text-decoration-resize`
* 🎉 `text-decoration-edge`
* ❤️ `text-decoration-spread`

And again, multiple votes are allowed! Please read the comments for context! Note that we [resolved on](https://github.com/w3c/csswg-drafts/issues/8402#issuecomment-3407283833) renaming `text-decoration-*` and the `{row|column}-rule-*` properties for adjusting their length to a common suffix. So when choosing your option please keep that in mind!

Sebastian

**Comment by SebastianZ** [comment]
@mirisuzanne I've now posted two new comments for the poll. I think it's ok now to share it.

Sebastian

**Comment by AlaskanEmily** [comment]
I'm currently working on this feature in Gecko.
`text-decoration-length-adjust`/`text-decoration-size-adjust` are very long.
I like `text-decoration-inset` as that indicates what positive/negative values would do.

**Comment by rianmurnen** [comment]
Consider describing this feature of a line as the “cap” per prior art in Adobe Illustrator, and probably even older graphics software.

`text-decoration-cap`

Here is an Adobe Illustrator help doc, [Apply a Stroke to an Object](https://helpx.adobe.com/illustrator/using/stroke-object.html?x-product=Helpx%2F1.0.0&x-product-location=Search%3AForums%3Alink%2F3.7.2-dev.2). Notice the initial photo of the Stroke Panel which offers a few basic options to adjust the cap style (butt, round, projecting). Also see the section “Change the caps or joins of a line”.

While the “cap” in Illustrator doesn’t afford arbitrary length values, the concept of describing this feature of a line could be helpful for CSS.

Example

- `text-decoration-cap: normal` (butt cap)
- `text-decoration-cap: round` (rounded cap)
- `text-decoration-cap: <length>` a positive or negative length resulting in either a extended/projecting cap (positive  value) or shortened cap (negative value)

**Comment by mirisuzanne** [comment]
I like the idea of having different cap shapes available for a decoration or rule, but I wouldn't expect that shape to cascade in the same property as a length/offset. Unless `-cap` is a shorthand for other `-cap-*` properties (eg `*-cap-inset` and `*-cap-shape`) in which case we're back to the question at hand.

**Comment by rianmurnen** [comment]
I'm missing something about the cascade I guess.

What made me think of Illustrators stroke caps was how the term “cap” is

- concise
- names the part of the object rather than action to take on the part
- allows for named preset values such as Illustrator's “projected” which extends the cap equal to the `text-decoration-thickness`. Might be sufficient for the most common case for extending the line?

Perhaps I need to go read the other threads to understand the context and use cases for extending an arbitrary length past the end of the text.

**Comment by mirisuzanne** [comment]
I agree that's useful context to add, and points to some interesting potential features.

> I'm missing something about the cascade I guess.

When two things are set in a single property you can't change one without overriding the other. This is true of shorthands generally - when you set `border: solid`, that resets `border-width` and `border-color` to their initial values. If we didn't have longhand properties for each part (as with shadows), there would be no way to adjust just the `border-width` while leaving everything else in place. By having separate properties, each part can _cascade independently_, without overriding the other parts.

So the question is: should you always have to set the 'cap' shape and `<length>` together in one property? Or should it be possible to change one without overriding the other? If we ever want to adjust those values separately, we need distinct properties for them to use. And then we would still need unique names for the cap shape and the cap… length? offset? something else?

I kinda like `*-cap-shape` and `*-cap-offset` (to match `text-underline-offset`), but it loses the brevity you were going for.

**Comment by rianmurnen** [comment]
@oddbird Yeah, the cap shape idea does open up the idea of custom start/end terminals like ball, square, arrowheads, and wanting size/scale them independent of `-thickness`.

I wasn't thinking about combing a shape with a value (round 10px), but I see what you mean about the cascade and shorthand in that approach.

**Comment by Crissov** [comment]
Just for the record, CSS already inherits the term _cap_ from SVG for the [`stroke-linecap` property](https://drafts.fxtf.org/fill-stroke-3/#propdef-stroke-linecap) in the Fill & Stroke Module. (The hyphen one would expect therein is missing in fact just because.)

For what it’s worth, CSS Fonts always abbreviates plural ”capitals“ as `caps` in its property names and keyword values, but the related unit in CSS Values is called `cap`, as is the respective keyword for `line-fit-edge` in CSS Inline.

**Comment by fantasai** [comment]
I'm a bit disappointed that the existing name `column-rule-outset` wasn't included in the poll. Though I suppose `text-decoration-outset` and `text-underline-offset` are maybe too close together? Still, would have been nice to see how people react to it.

I'm also still a bit concerned that `inset`, as @jfkthame mentioned, is associated with layout properties that move things around (rather than trimming them)...

**Comment by jfkthame** [comment]
In my opinion, `-outset` has the same drawback as `-inset`, that it sounds like it could be affecting layout and not merely rendering; but in addition, it would have us using negative values for the common case, at least in the case of text-decoration (I'm not sure which way the balance is likely to lie for column-rule). I also think "outset" is less familiar in general as an everyday word/concept than "inset".

While I personally still like `-trim`, I think the overall preference for `-inset` seen in https://github.com/w3c/csswg-drafts/issues/8402#issuecomment-3408427694 is pretty clear, so I'm suggesting we should resolve on that; I'm not sure there's much more useful discussion to be had at this point.

**Comment by mirisuzanne** [comment]
`inset()` is also a basic shape, which is used for rendering concerns, not for layout. I disagree that these terms are somehow layout-specific.

**Comment by SebastianZ** [comment]
The CSS Text Decorations part just got merged. #13024 targets the CSS Gap Decorations part.

Sebastian

**Comment by jav099** [comment]
The CSS Gap Decorations part just merged, should this now be marked as closed?

**Comment by SebastianZ** [comment]
Yes, I think so. Thanks for the change, @jav099!

Sebastian
